File: frontend/src/api/clinic.ts
import { apiFetch, apiFetchBlob } from "./client";

export interface ErrorCodeItem {
  code: string;
  display: string;
}

export interface QuestionSearchResult {
  question_id: number;
  school: string | null;
  year: number | null;
  semester: number | null;
  exam_type: string | null;
  question_number: number;
  chapter: string | null;
  difficulty: string | null;
  text_preview: string;
}

export type SourceMode = "db" | "external";

export interface ClinicEntryRequest {
  student_id: number;
  source_mode: SourceMode;
  wrong_question_id?: number | null;
  external_label?: string | null;
  error_code: string;
  keyword: string;
  wrong_date: string;
}

export interface RetrySchedule {
  d3: string;
  d7: string;
  d14: string;
  d30: string;
}

export interface ClinicEntryResponse {
  entry_id: number;
  mode: SourceMode;
  prescribed_qids: number[];
  prescription_question_ids: number[];
  schedule: RetrySchedule;
  student_name: string;
  error_code: string;
  external_label: string | null;
}

export interface PendingRetry {
  entry_id: number;
  student_id: number;
  student_name: string;
  wrong_question_id: number | null;
  external_label: string | null;
  wrong_date: string;
  error_code: string;
  due_flags: string[];
}

export function fetchClinicMeta(): Promise<{ error_codes: ErrorCodeItem[] }> {
  return apiFetch("/api/clinic/meta");
}

export function searchClinicQuestions(keyword: string): Promise<QuestionSearchResult[]> {
  return apiFetch(`/api/clinic/search-questions?keyword=${encodeURIComponent(keyword)}`);
}

export function createClinicEntry(req: ClinicEntryRequest): Promise<ClinicEntryResponse> {
  return apiFetch<ClinicEntryResponse>("/api/clinic/entries", {
    method: "POST",
    body: JSON.stringify(req),
  });
}

export function downloadPrescriptionPdf(
  question_ids: number[], title: string, subtitle: string,
): Promise<Blob> {
  return apiFetchBlob("/api/clinic/prescription-pdf", {
    method: "POST",
    body: JSON.stringify({ question_ids, title, subtitle }),
  });
}

export function fetchPendingRetries(): Promise<PendingRetry[]> {
  return apiFetch("/api/clinic/pending-retries");
}

File: backend/app/schemas/clinic.py
from __future__ import annotations

from typing import Literal, Optional

from pydantic import BaseModel, Field


class ErrorCodeItem(BaseModel):
    code: str
    display: str


class ClinicMetaResponse(BaseModel):
    error_codes: list[ErrorCodeItem]


class QuestionSearchResult(BaseModel):
    question_id: int
    school: Optional[str] = None
    year: Optional[int] = None
    semester: Optional[int] = None
    exam_type: Optional[str] = None
    question_number: int
    chapter: Optional[str] = None
    difficulty: Optional[str] = None
    text_preview: str


class ClinicEntryRequest(BaseModel):
    student_id: int
    source_mode: Literal["db", "external"]
    wrong_question_id: Optional[int] = None
    external_label: Optional[str] = None
    error_code: str
    keyword: str = ""
    wrong_date: str


class RetrySchedule(BaseModel):
    d3: str
    d7: str
    d14: str
    d30: str


class ClinicEntryResponse(BaseModel):
    entry_id: int
    mode: Literal["db", "external"]
    prescribed_qids: list[int] = []
    prescription_question_ids: list[int] = []
    schedule: RetrySchedule
    student_name: str
    error_code: str
    external_label: Optional[str] = None


class PrescriptionPdfRequest(BaseModel):
    question_ids: list[int]
    title: str
    subtitle: str


class PendingRetry(BaseModel):
    entry_id: int
    student_id: int
    student_name: str
    wrong_question_id: Optional[int] = None
    external_label: Optional[str] = None
    wrong_date: str
    error_code: str
    due_flags: list[str]
